### PR TITLE
kepify: New command to generate KEP json for Netlify

### PR DESCRIPTION
**Pull request by dims**
`go run cmd/kepify/main.go --dir keps --output out.json`

### CODE REVIEW

**Comment by dims** [comment]
cc @saschagrunert @jeefy 

here's a crack at what we talked about on friday. running the command generates a json file which hopefully can be massaged to work with netlify. PTAL

File: cmd/kepify/main.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "authors", strings.Join(kep.Authors, ","))
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "editor", kep.Editor)
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "creation-date", kep.CreationDate)
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "last-updated", kep.LastUpdated)
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "status", kep.Status)
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "see-also", strings.Join(kep.SeeAlso, ","))
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "replaces", strings.Join(kep.Replaces, ","))
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "superseded-by", strings.Join(kep.SupersededBy, ","))
+		contents, _ := json.Marshal(kep.Contents)
+		fmt.Fprintf(file, "\t\t\"%s\": %s\n", "markdown", contents)
```
Comment (saschagrunert) [inline cmd/kepify/main.go]: Hm, can we use the build tags to serialize the JSON like we already do for the YAML part?
https://github.com/kubernetes/enhancements/blob/502b4a7ace29104780658b41527a10ee9806b3c7/pkg/kepval/keps/proposals.go#L37-L49

File: cmd/kepify/main.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+		return err
+	}
+	defer file.Close()
+
+	total := len(proposals)
+	fmt.Printf("Total KEPs: %d\n", total)
+
+	fmt.Fprintln(file, "{")
+	for i, kep := range proposals {
+		fmt.Fprintf(file, "\t\"%s\": {\n", hash(kep.OwningSIG+":"+kep.Title))
```
Comment (saschagrunert) [inline cmd/kepify/main.go]: I think this is okay, but makes the key unusable for consumers. We probably want to switch to a plain array and remove the key entirely. Or we use the KEPs path in the repository as key, this way we could easily link it back into this repository.

File: cmd/kepify/main.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "title", kep.Title)
```
Comment (saschagrunert) [inline cmd/kepify/main.go]: From the example JSON I see that the titles differ in its style: like `aws-k8s-tester` vs `Custom endpoints support for AWS Cloud Provider`. Do you think we can avoid that somehow?

File: cmd/kepify/main.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "owning-sig", kep.OwningSIG)
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "participating-sigs", strings.Join(kep.ParticipatingSIGs, ","))
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "reviewers", strings.Join(kep.Reviewers, ","))
```
Comment (saschagrunert) [inline cmd/kepify/main.go]: It would be good if this will be a plain JSON array.

File: cmd/kepify/main.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	defer file.Close()
+
+	total := len(proposals)
+	fmt.Printf("Total KEPs: %d\n", total)
+
+	fmt.Fprintln(file, "{")
+	for i, kep := range proposals {
+		fmt.Fprintf(file, "\t\"%s\": {\n", hash(kep.OwningSIG+":"+kep.Title))
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "title", kep.Title)
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "owning-sig", kep.OwningSIG)
```
Comment (saschagrunert) [inline cmd/kepify/main.go]: Stripping the `sig-` prefix would probably help. The same applies for the participating SIGs.

File: cmd/kepify/main.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "participating-sigs", strings.Join(kep.ParticipatingSIGs, ","))
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "reviewers", strings.Join(kep.Reviewers, ","))
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "authors", strings.Join(kep.Authors, ","))
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "editor", kep.Editor)
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "creation-date", kep.CreationDate)
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "last-updated", kep.LastUpdated)
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "status", kep.Status)
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "see-also", strings.Join(kep.SeeAlso, ","))
+		fmt.Fprintf(file, "\t\t\"%s\": \"%s\",\n", "replaces", strings.Join(kep.Replaces, ","))
```
Comment (saschagrunert) [inline cmd/kepify/main.go]: If the fields are empty, then I suggest in removing them or setting them to `null`.

File: cmd/kepify/main.go (hunk 1 above)
Comment (saschagrunert) [inline cmd/kepify/main.go]: I think we could somehow split the table of contents from the markdown. This way we easily could preview the KEP only by its TOC.

**Comment by saschagrunert** [comment]
Thank you for working on this :pray: I just have a couple of comments.

File: cmd/kepify/main.go (hunk 1 above)
Comment (dims) [reply cmd/kepify/main.go]: @saschagrunert this gives us more control over the output if we run into any netlify idiosynchracies. we can always do this in a follow up PR.

File: cmd/kepify/main.go (hunk 2 above)
Comment (dims) [reply cmd/kepify/main.go]: the hash of sig + title is unique, there was discussion about moving the keps around (separate directories for keps in different status etc), so i needed something that will stay across any such moves. note that we have the `Filename` in the data structure as well, So worst case we just ignore this hash.

Comment (saschagrunert) [reply cmd/kepify/main.go]: Makes sense :+1:

File: cmd/kepify/main.go (hunk 3 above)
Comment (dims) [reply cmd/kepify/main.go]: not in this PR, we have to raise this as a problem in https://github.com/kubernetes/enhancements/issues/822

File: cmd/kepify/main.go (hunk 4 above)
Comment (dims) [reply cmd/kepify/main.go]: DONE

File: cmd/kepify/main.go (hunk 5 above)
Comment (dims) [reply cmd/kepify/main.go]: Can we do this in netlify? i'd like the markdown to have the "official" names

File: cmd/kepify/main.go (hunk 1 above)
Comment (dims) [reply cmd/kepify/main.go]: let's do this in a follow up please

File: cmd/kepify/main.go (hunk 5 above)
Comment (saschagrunert) [reply cmd/kepify/main.go]: Yes, sure.

File: cmd/kepify/main.go (hunk 6 above)
Comment (dims) [reply cmd/kepify/main.go]: DONE (see sample output in https://gist.github.com/dims/97ca19d97d74c5ec697f7d01c5dc1247)

File: cmd/kepify/main.go (hunk 1 above)
Comment (saschagrunert) [reply cmd/kepify/main.go]: Okay, good :+1:

**Comment by dims** [comment]
updated generated json in https://gist.github.com/dims/97ca19d97d74c5ec697f7d01c5dc1247

**Comment by dims** [comment]
/assign @justaugustus @jdumars @calebamiles 

Folks, let's please merge and iterate, this is just the first cut. Hoping @saschagrunert @jeefy and others can help tweak things as we try this with netlify

**Comment by dims** [comment]
thanks @justaugustus

**Comment by saschagrunert** [comment]
@dims a first implementation proposal can be found here: https://github.com/kubernetes-sigs/release-notes/pull/140

We probably want to setup a new k-sigs project for this.

**Comment by dims** [comment]
@saschagrunert new repo sounds good! please request one at https://github.com/kubernetes/org/issues

**Pull request merged**